app/services/update_location.rb

```diff
@@ -6,30 +6,45 @@ def initialize(location:, user:)
     @user = user
   end
 
-  def update!(params)
-    params = params.with_indifferent_access
-    if params['address']
-      params['address_id'] = Address.find_or_create_from_params(params.delete('address')).id
+  def update(params)
+    address = params.delete('address')
+    if address
+      address_params = build_address(address)
+      params.merge!(address_params)
     end
 
     return location if location.matches_params?(params)
 
     Location.transaction do
-      location.update_attributes!(state: 'old')
-      create_new_version!(params)
+      begin
+        location.update_attributes!(state: 'old')
+        create_new_version(params)
+      rescue ActiveRecord::RecordInvalid
```

Comment:
Ah, so you don't want the exception to propogate outside of the transaction scope, I get it. It's a strange pattern because on seeing the `raise` you naturally expect it to propogate but transaction scopes have special properties in that regard.